### ISSUE TITLE
Assline data sticks are not updating

### ISSUE THREAD

**Issue by GTNH-Colen** [GTNewHorizons/GT-New-Horizons-Modpack#8476]
### Your GTNH Discord Username

Colen (His Supreme Holiness, HBM manager XIV)

### Your Pack Version

2.1.1.0

### Your Expectation

Essentially data sticks store their recipe internally and then the assline uses that as gospel. If I make a data stick that says 1 dirt = 1 LuV motor then the assline will follow that over the true intended recipe from NEI.

This is problematic given how we do quite often change recipes around, there are data sticks on some of the officials which are *years* out of date and still function.

### Your Proposal

Fix it (please uwu).

I believe the assline code is quite messy but there are probably some workarounds we could do just to make sure that a data stick is using the latest intended version?

### Final Checklist

- [X] I have searched this issue tracker and there is nothing similar already. Posting on a closed issue saying the bug still exists will prompt us to investigate and reopen it once we confirm your report.
- [X] I can reproduce this problem consistently by follow the exact steps I described above, or this does not need reproducing, e.g. recipe loophole.
- [X] I have asked other people and they confirm they also have this problem by follow the exact steps I described above, or this does not need reproducing, e.g. recipe loophole.

**Comment by yukieiji** [comment]
I think it would be difficult to go through the data sticks that exist in the world and check if they all belong to AssLine's resipes.
So, how about adding a machine that will check if the inserted data stick matches the NEI recipe and update it if it doesn't?

**Comment by Prometheus0000** [comment]
The idea is not to store the recipe on the data stick, but just make the data stick say 'needed for making item x', which allows making item x, and then get the info from the recipe in the code like normal crafts.

No clue how to do that though.

**Comment by GlodBlock** [comment]
directly read from data stick will be more fast and efficient than search in the whole recipe collection

**Comment by Prometheus0000** [comment]
Can it be made to update on load or something though?

**Comment by leagris** [comment]
Technically data-sticks could store a recipe hash that would be used to revoke a recipe when the stick is used; but before we go further on this, I'd like to figure the effect of this issue.

What I see, is those who invested effort into their ASSLine and accompanying logistic, don't have to change it all as long as ingredients are available. This look like a better situation than with AE2 patterns that turns to junk at any upcoming upgrade without even getting user prior notification.

What kind of negative impact has the current data stick system?

A little while ago I have discussed in Discord the possibility for player to gain better data-stick recipes, like more output, less material, better efficiency. These enhanced data-sticks would take of the advantage, the recipe is written on it.

- These enhanced recipes data-sticks could be quest rewards
- These could be researched some way.
- These could be rare boss loot.

To sum it up, not sure it is worth changing these data sticks

**Comment by Prometheus0000** [comment]
The idea is to invalidate old recipes for when/if they get changed. So you don't allow people to use the old recipe, that may be easier than the new one. Though I would hope that they don't constantly become harder of course.

I do like the idea of better recipes though. But if you put them only on the stick and not in NEI, they become hard to figure out again.

**Comment by GlodBlock** [comment]
we can add version info data stick and check it in `checkRecipe(ItemStack aStack)`. e.g. if in the 2.1.1.0 pack and the assline find the data stick's version info is 2.0.8.0, it will search the current recipe collection and update it. since the version update will only be executed at most once, it won't influence the performance too much.

as to the special designed data stick, add a special tag on it and assline won't check its version info in checkRecipe.

**Comment by GTNH-Colen** [comment]
I think @GlodBlock made the best suggestion here. I would love to see cheaper researchable recipes for high computational power from the quantum computer so leaving the possibility of that open is brilliant.

However I do take issue with extremely old and unbalanced recipes still being usable, I understand the frustration of have a recipe made invalid occasionally but this totally bypasses intended mechanics and is a fairly serious issue when balance changes are made.

Eg there are legacy nano circuit data sticks on some of the officials which are vastly cheaper than their modern day counterparts.

**Comment by GTNH-Colen** [comment]
Maybe we should do this before next update? 🥰 bug major after all.

**Comment by boubou19** [comment]
It's not a bug as the recipes are hardcoded in the stick with the scanner or with the research station. Then the assembly line makes recipes based on the data stored in the stick. It's a design choice that prevents the assembly lines automation with the update. It's both good and bad at the same time: asslines are complex automations and getting them broken upon pack update is harsh, but if the recipe was nerfed then players will likely keep the old recipe if they think the nerf isn't making any sense. But if the recipes are buffed, then players will gladly update the sticks on their own.

Now if we want to make a change, why not, but it's not a bug :)

**Comment by GTNH-Colen** [comment]
You can't gaslight your way out of this one, it's a blatant bug that only becomes more exploitable each update 😂 

It's obviously unintended that people can continue to use old recipes. There is no arguing your way out of that.

**Comment by boubou19** [comment]
Icebox tag says it all :D.

More seriously if this wasn't intended then why was it designed that way?

**Comment by Sphyix** [comment]
Clearly intended/suggestion of a change.

**Comment by GTNH-Colen** [comment]
> More seriously if this wasn't intended then why was it designed that way? 

I don't think Greg or whoever made the assline had an ever changing pack spanning 5+ years in mind when he made it.

**Comment by Sphyix** [comment]
I don't want to redo all the research and all the setups everytime a recipe is changed.
And colen, go back to lag eta :)

**Comment by GTNH-Colen** [comment]
Assline recipes are not changed that much but I don't think people should be allowed to use old recipes completely bypassing balance just because they don't feel like updating a setup. In no other area would this be acceptable and given everyone is using Alex's assline design it's very easy to update it anyways.

This isn't some minor thing, there are huge exploits because of old assline sticks. Ultra cheap LuV motors, completely bypassing UHV circuits for nanos etc.

**Comment by leagris** [comment]
Solution:

1. Replace direct ingredients references written in memory stick, by a single reference ID to the recipe stored in the memory stick.
2. Invalidate all old-style stored ingredients memory sticks.

Alternate solution:

1. Change assembly-line to only use the output item stack as a reference pointer to the recipe.

Alternate solution works even without invalidating existing memory sticks.

**Comment by GTNH-Colen** [comment]
First solution sounds good.

> Alternate solution:
> 
> 1. Change assembly-line to only use the output item stack as a reference pointer to the recipe.
> 
> Alternate solution works even without invalidating existing memory sticks.

This should not be an option, it is an *exploit* plain and simple. People may enjoy using it but having custom recipes that are being saved is inherently broken. If I have a data stick that says 1 cobble = 1 UEV circuit, is that not broken because the stick says so?

**Comment by leagris** [comment]
> First solution sounds good.
> 
> > Alternate solution:
> > 
> > 1. Change assembly-line to only use the output item stack as a reference pointer to the recipe.
> > 
> > Alternate solution works even without invalidating existing memory sticks.
> 
> This should not be an option, it is an _exploit_ plain and simple. People may enjoy using it but having custom recipes that are being saved is inherently broken. If I have a data stick that says 1 cobble = 1 UEV circuit, is that not broken because the stick says so?

2nd solution does not invalidate existing memory sticks but cause the old memory sticks to use the output item stack as a pointer to the actual recipe. All ingredients stored in old memory sticks would be just ignored.

**Comment by draknyte1** [comment]
Assign it to me and I’ll fix this dumb shit. It’s not hard. 🤨
![91DF4529-AA4C-4C0D-BAC4-BEAA10C2579C](https://user-images.githubusercontent.com/3060479/141859099-62e2a42e-e547-458a-803c-b4db23d46aba.jpeg)

**Comment by bombcar** [comment]
Fixed by https://github.com/GTNewHorizons/GT5-Unofficial/pull/735 I believe
